### PR TITLE
1157 fix multiple hashing and experiment initialisation

### PR DESCRIPTION
**Pull request by cmarooney-stfc**
As per point 1 of #1157, objects are hashed twice, unnecessarily, for unique_object_containers. This does not happen for these containers themselves, but when used in unique_references_containers, this does happen, and is fixed by this PR.

This investigation also shows that sqw had its experiment_info incorrectly initialised by a property definition rather than a part of the constructor. This may also fix point 2 about clearing the container although as there are no instructions for reproducing it I cannot be certain.

It also shows that unique_references_containers are misleadingly tested by the present tests, as earlier tests perform initialisation of the persistent data which then allow later tests to proceed which would otherwise fail if run in isolation.

### CODE REVIEW

File: horace_core/sqw/@sqw/sqw.m [hunk 1]
```diff
@@ -65,7 +65,7 @@
         % The class providing brief description of a whole sqw file.
         main_header_ = main_header_cl();
 
-        experiment_info_ = Experiment();
+        experiment_info_ = []; %Experiment();
```
Comment (oerc0122) [inline horace_core/sqw/@sqw/sqw.m:68]: So the array has the right type.
```suggestion
        experiment_info_ = Experiment.empty();
```

File: herbert_core/utilities/classes/@unique_references_container/unique_references_container.m [hunk 2]
```diff
@@ -498,10 +498,10 @@
...
                     'global name unset');
             end
-            [glindex, ~] = self.global_container('value',self.global_name_).find_in_container(inobj);
+            [glindex, hash] = self.global_container('value',self.global_name_).find_in_container(inobj);
             if isempty(glindex)
                 glcont = self.global_container('value',self.global_name_);
-                [glcont,glindex] = glcont.add(inobj);
+                [glcont,glindex] = glcont.add_single_(inobj,[],hash);
                 if glindex == 0
                     % object was not added
```
Comment (oerc0122) [inline herbert_core/utilities/classes/@unique_references_container/unique_references_container.m:506]: What is the "standard" signal that an object has not been added? Reading this is appears that empty `glindex` or `glindex == 0`, but that is two different signals. Is this correct?

**Review by oerc0122 — CHANGES_REQUESTED** [review]

**Review by abuts — COMMENTED** [review]
This is a comment:
As object hash is caclulated in `find_in_container` I would assume that external hash should be propagated up to `fild_in_container` to return if provided. 

This looks like clearer code and less problem for maintaining it. 

Bearing in mind idea of storing hash with object, current change addresses local problem so fine.

File: herbert_core/utilities/classes/@unique_objects_container/private/add_single_.m
Comment (tgperring) [inline herbert_core/utilities/classes/@unique_objects_container/private/add_single_.m:1]: I am not competent to give a review without spending time familiarising myself with the structure of the problem.

File: horace_core/sqw/@sqw/sqw.m (hunk 1 above)
Comment (cmarooney-stfc) [reply horace_core/sqw/@sqw/sqw.m:68]: But I don't want it to have the right type. I want it to fail as obviously as possible if this variable has not been reinitialised correctly, probably in the constructor.

File: herbert_core/utilities/classes/@unique_references_container/unique_references_container.m (hunk 2 above)
Comment (cmarooney-stfc) [reply herbert_core/utilities/classes/@unique_references_container/unique_references_container.m:506]: Very well spotted - it's an issue - but what is happening is that if isempty(glindex), the object has not been found and is added as a new object. If glindex==0 then nothing has been added to the container by add_single_, so the call has failed - not the same as not being found. However, a look at add_single_ shows that an object will always be added, so the if glindex==0 block is paranoia, and should be removed - I will do so.

File: herbert_core/utilities/classes/@unique_objects_container/private/add_single_.m [hunk 3]
```diff
@@ -1,12 +1,29 @@
...
+%        this will have been found from a previous call of find_in_container
+% hash - the hash for obj previously made by that call to find_in_container
+% 
+% Output
+% ------
+% self - the modified container (modified by adding obj)
+% nuix - the insertion index at which obj is added in the container
+
+% if ix and hash are not specified, call find_in_container to get them
+if nargin<=2
```
Comment (oerc0122) [inline herbert_core/utilities/classes/@unique_objects_container/private/add_single_.m:24]: Suggestion:
I prefer not using `nargin` to check arguments as it limits future function signature changes. Instead, prefer explicitly checking `exist('variable', 'var')`

**Review by oerc0122 — APPROVED** [review]

**Pull request merged**